capa/features/extractors/ghidra/file.py

```diff
@@ -0,0 +1,221 @@
+#  is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and limitations under the License.
+
+import struct
+from typing import Tuple, Iterator
+
+import ghidra
+from ghidra.program.flatapi import FlatProgramAPI
+from ghidra.program.model.symbol import SourceType, SymbolType
+
+import capa.features.extractors.common
+import capa.features.extractors.helpers
+import capa.features.extractors.strings
+import capa.features.extractors.ghidra.helpers
+from capa.features.file import Export, Import, Section, FunctionName
+from capa.features.common import FORMAT_PE, FORMAT_ELF, Format, String, Feature, Characteristic
+from capa.features.address import NO_ADDRESS, Address, FileOffsetAddress, AbsoluteVirtualAddress
+
+currentProgram: ghidra.program.database.ProgramDB
+flatapi = FlatProgramAPI(currentProgram)
+
+MAX_OFFSET_PE_AFTER_MZ = 0x200
+
+
+def check_segment_for_pe() -> Iterator[Tuple[int, int]]:
+    """check segment for embedded PE
+
+    adapted for Ghidra from:
+    https://github.com/vivisect/vivisect/blob/91e8419a861f4977https://github.com/vivisect/vivisect/blob/91e8419a861f49779f18316f155311967e696836/PE/carve.py#L259f18316f155311967e696836/PE/carve.py#L25
+    """
+
+    mz_xor = [
+        (
+            capa.features.extractors.helpers.xor_static(b"MZ", i),
+            capa.features.extractors.helpers.xor_static(b"PE", i),
+            i,
+        )
+        for i in range(256)
+    ]
+
+    todo = []
+    for mzx, pex, i in mz_xor:
+        # find all segment offsets containing XOR'd "MZ" bytes
+        off: ghidra.program.model.address.GenericAddress
+        for off in capa.features.extractors.ghidra.helpers.find_byte_sequence(mzx):
+            todo.append((off, mzx, pex, i))
+
+    seg_max = currentProgram.getMaxAddress()
+    while len(todo):
+        off, mzx, pex, i = todo.pop()
+
+        # MZ header has one field we will check e_lfanew is at 0x3c
+        e_lfanew = off.add(0x3C)
+
+        if seg_max.getOffset() < (e_lfanew.getOffset() + 4):
+            continue
+
+        e_lfanew_bytes = b""
+        try:
+            e_lfanew_sbytes = flatapi.getBytes(e_lfanew, 4)
+            for b in e_lfanew_sbytes:
+                b = (b & 0xFF).to_bytes(1, "little")
+                e_lfanew_bytes = e_lfanew_bytes + b
+        except RuntimeError:  # no bytes will be returned, so we can bail out here
+            return
+
+        newoff = struct.unpack("<I", capa.features.extractors.helpers.xor_static(e_lfanew_bytes, i))[0]
+
+        # assume XOR'd "PE" bytes exist within threshold
+        if newoff > MAX_OFFSET_PE_AFTER_MZ:
+            continue
+
+        peoff = off.add(newoff)
+        if seg_max.getOffset() < (peoff.getOffset() + 2):
+            continue
+
+        pe_bytes = b""
+        try:
+            pe_off_bytes = flatapi.getBytes(peoff, 2)
```

Comment:
Let’s make these lines into a single helper function that we can call elsewhere.